ql/src/java/org/apache/hadoop/hive/ql/io/parquet/vector/VectorizedParquetRecordReader.java

```diff
@@ -0,0 +1,658 @@
+   * @return The Type object of the field found; null otherwise.
+   */
+  private static Type getFieldTypeIgnoreCase(GroupType groupType, String fieldName) {
+    for (Type type : groupType.getFields()) {
+      if (type.getName().equalsIgnoreCase(fieldName)) {
+        return type;
+      }
+    }
+
+    return null;
+  }
+
+
+  /**
+   * Searchs column names by name on a given Parquet message schema, and returns its projected
+   * Parquet schema types.
+   *
+   * @param schema Message type schema where to search for column names.
+   * @param colNames List of column names.
+   * @param colTypes List of column types.
+   * @return A MessageType object of projected columns.
+   */
+  private static MessageType getSchemaByName(MessageType schema, List<String> colNames, List<TypeInfo> colTypes) {
+    List<Type> projectedFields = getProjectedGroupFields(schema, colNames, colTypes);
+    Type[] typesArray = projectedFields.toArray(new Type[0]);
+
+    return Types.buildMessage()
+      .addFields(typesArray)
+      .named(schema.getName());
+  }
+
+  /**
+   * Searchs column names by index on a given Parquet file schema, and returns its corresponded
+   * Parquet schema types.
+   *
+   * @param schema Message schema where to search for column names.
+   * @param colNames List of column names.
+   * @param colIndexes List of column indexes.
+   * @return A MessageType object of the column names found.
+   */
+  private static MessageType getSchemaByIndex(MessageType schema, List<String> colNames, List<Integer> colIndexes) {
+    List<Type> schemaTypes = new ArrayList<Type>();
+
+    for (Integer i : colIndexes) {
+      if (i < colNames.size()) {
+        if (i < schema.getFieldCount()) {
+          schemaTypes.add(schema.getType(i));
+        } else {
+          //prefixing with '_mask_' to ensure no conflict with named
+          //columns in the file schema
+          schemaTypes.add(Types.optional(PrimitiveType.PrimitiveTypeName.BINARY).named("_mask_" + colNames.get(i)));
+        }
+      }
+    }
+    return new MessageType(schema.getName(), schemaTypes);
+  }
+
+  /**
+   * columnBatch object that is used for batch decoding. This is created on first use and triggers
+   * batched decoding. It is not valid to interleave calls to the batched interface with the row
+   * by row RecordReader APIs.
+   * This is only enabled with additional flags for development. This is still a work in progress
+   * and currently unsupported cases will fail with potentially difficult to diagnose errors.
+   * This should be only turned on for development to work on this feature.
+   *
+   * When this is set, the code will branch early on in the RecordReader APIs. There is no shared
+   * code between the path that uses the MR decoders and the vectorized ones.
+   *
+   * TODOs:
+   *  - Implement v2 page formats (just make sure we create the correct decoders).
+   */
+  private VectorizedRowBatch columnarBatch;
```

Comment:
Not sure if we need this field. The batch is passed in in `next` method. For `createValue` we can just create a new batch and return it.